internal/oc/release.go

```diff
@@ -442,21 +444,158 @@ func execute(log logrus.FieldLogger, executer executer.Executer, pullSecret stri
+		return
+	}
+
+	mirrorsRegistriesConfig, err := r.mirrorRegistriesBuilder.ExtractLocationMirrorDataFromRegistries()
+	if err != nil {
+		log.WithError(err).Errorf("Failed to get the mirror registries needed for mirrors file")
+		return
+	}
+
+	// Check if we should use the deprecated ImageContentPolicySource object or the new ImageDigestMirrorSet. To do
+	// so we check if the 'oc adm release info' command supports the '--idms-file' flag.
+	suportsIdmsFileFlag, err := r.supportsIdmsFileFlag(log, templateReleaseInfo)
+	if err != nil {
+		return
+	}
+	var (
+		name string
+		file string
+	)
+	if suportsIdmsFileFlag {
+		name = idmsFileFlagName
+		file, err = r.getIdmsFileFromRegistriesConfig(log, mirrorsRegistriesConfig)
+	} else {
+		name = icspFileFlagName
+		file, err = r.getIcspFileFromRegistriesConfig(log, mirrorsRegistriesConfig)
+	}
+	if err != nil {
+		return
+	}
+	result = &mirrorsFlagInfo{
+		logger: log,
+		name:   name,
+		file:   file,
+	}
+	return
+}
+
+// mirrorsFlagInfo contains the information needed to populate the '--idms-file' or '--icsp-file' command line flag that
+// is used by the 'oc' command to specify the mirrors configuration.
+type mirrorsFlagInfo struct {
+	// logger is the logger used by the methods of this type.
+	logger logrus.FieldLogger
+
+	// name is the name of the command line flag, 'idms-file' or 'icsp-file'.
+	name string
+
+	// file is the name of a temporary file containing the serialized ImageDigestMirrorSet or
+	// ImageContentSourcePolicy.
+	file string
+}
+
+// String returns the complete text of the flag, something like '--idms-file=/tmp/my-idms.yaml'.
+func (i *mirrorsFlagInfo) String() string {
+	if i == nil {
+		return ""
+	}
+	return fmt.Sprintf("--%s=%s", i.name, i.file)
+}
+
+// Delete deletes the temporary file.
+func (i *mirrorsFlagInfo) Delete() {
+	if i == nil {
+		return
+	}
+	err := os.Remove(i.file)
+	if err != nil {
+		i.logger.WithError(err).WithFields(logrus.Fields{
+			"file": i.file,
+		}).Error("Failed to delete temporary file")
+	}
+}
+
+// Create a temporary file containing the ImageDigestMirrorSet
+func (r *release) getIdmsFileFromRegistriesConfig(log logrus.FieldLogger,
+	mirrorRegistriesConfig []mirrorregistries.RegistriesConf) (result string, err error) {
+	contents, err := getIdmsContents(mirrorRegistriesConfig)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to create the IDMS file from registries.conf")
+		return "", err
+	}
+	if contents == nil {
+		log.Debugf("No registry entries to build IDMS file")
 		return "", nil
 	}
 
-	mirrorRegistriesConfig, err := r.mirrorRegistriesBuilder.ExtractLocationMirrorDataFromRegistries()
+	idmsFile, err := os.CreateTemp("", "idms-file")
 	if err != nil {
-		log.WithError(err).Errorf("Failed to get the mirror registries needed for ImageContentSources")
 		return "", err
 	}
+	log.Debugf("Building IDMS file from registries.conf with contents %s", contents)
+	if _, err := idmsFile.Write(contents); err != nil {
+		idmsFile.Close()
```

Comment:
What do you think about doing this in the function body?
```
defer idmsFile.Close()
```

Reply:
Yes, makes sense. I tried to not change the existing "getIcspFileFromRegistriesConfig" function, and this is just a copy. I will change it.

Reply:
Actually after reading this more carefully I think that doing `defer idmsFile.Close()` will not work correctly in all cases: if writing to the file fails it would then try to close the file after removing it, which would fail.

Reply:
We could do something like this if we wanted to avoid repeating the `Close` call:

```diff
diff --git a/internal/oc/release.go b/internal/oc/release.go
index a37a6d7b1..b50bb113d 100644
--- a/internal/oc/release.go
+++ b/internal/oc/release.go
@@ -546,13 +546,17 @@ func (r *release) getIdmsFileFromRegistriesConfig(log logrus.FieldLogger,
        if err != nil {
                return "", err
        }
+       defer func() {
+               _ = idmsFile.Close()
+               if err != nil {
+                       _ = os.Remove(idmsFile.Name())
+                       result = ""
+               }
+       }()
        log.Debugf("Building IDMS file from registries.conf with contents %s", contents)
        if _, err := idmsFile.Write(contents); err != nil {
-               idmsFile.Close()
-               os.Remove(idmsFile.Name())
                return "", err
        }
-       idmsFile.Close()
 
        return idmsFile.Name(), nil
 }
```

But that is probably more obscure than the current code.

Reply:
Oh I see, thanks for looking into it! The original way you had it looks good then